src/charm_state.py

```diff
@@ -0,0 +1,124 @@
+    port: int = Field(None, ge=1, le=65536)
+    user: Optional[str]
+    password: Optional[str]
+    auth_type: Optional[smtp.AuthType]
+    transport_security: Optional[smtp.TransportSecurity]
+    domain: Optional[str]
+
+
+class CharmConfigInvalidError(Exception):
+    """Exception raised when a charm configuration is found to be invalid.
+
+    Attrs:
+        msg (str): Explanation of the error.
+    """
+
+    def __init__(self, msg: str):
+        """Initialize a new instance of the CharmConfigInvalidError exception.
+
+        Args:
+            msg (str): Explanation of the error.
+        """
+        self.msg = msg
+
+
+@dataclass
+class CharmState:  # pylint: disable=too-many-instance-attributes
+    """Represents the state of the SMTP Integrator charm.
+
+    Attrs:
+        host: The hostname or IP address of the outgoing SMTP relay.
+        port: The port of the outgoing SMTP relay.
+        user: The SMTP AUTH user to use for the outgoing SMTP relay.
+        password: The SMTP AUTH password to use for the outgoing SMTP relay.
+        auth_type: The type used to authenticate with the SMTP relay.
+        transport_security: The security protocol to use for the outgoing SMTP relay.
+        domain: The domain used by the sent emails from SMTP relay.
+    """
+
+    host: str
+    port: int
+    user: Optional[str]
+    password: Optional[str]
+    auth_type: Optional[smtp.AuthType]
+    transport_security: Optional[smtp.TransportSecurity]
+    domain: Optional[str]
+
+    def __init__(self, *, smtp_integrator_config: SmtpIntegratorConfig):
+        """Initialize a new instance of the CharmState class.
+
+        Args:
+            smtp_integrator_config: SMTP Integrator configuration.
+        """
+        self.host = smtp_integrator_config.host
+        self.port = smtp_integrator_config.port
+        self.user = smtp_integrator_config.user
+        self.password = smtp_integrator_config.password
+        self.auth_type = smtp_integrator_config.auth_type
+        self.transport_security = smtp_integrator_config.transport_security
+        self.domain = smtp_integrator_config.domain
+
+    @classmethod
+    def from_charm(cls, charm: "ops.CharmBase") -> "CharmState":
+        """Initialize a new instance of the CharmState class from the associated charm.
+
+        Args:
+            charm: The charm instance associated with this state.
+
+        Return:
+            The CharmState instance created by the provided charm.
+
+        Raises:
+            CharmConfigInvalidError: if the charm configuration is invalid.
+        """
+        config = {k: v for k, v in charm.config.items() if k in KNOWN_CHARM_CONFIG}
+        try:
+            # Incompatible with pydantic.AnyHttpUrl
+            valid_config = SmtpIntegratorConfig(**config)  # type: ignore
+        except ValidationError as exc:
+            error_fields = set(
+                itertools.chain.from_iterable(error["loc"] for error in exc.errors())
```

Comment:
Out of curiosity, what would be an example value of `error["loc"]` in this case please?

Reply:
`loc` is, per the pydantic docs, the error location, that is, the key whose value does not pass validation, i.e. `host`